content/blog/2019-09-25-function-calls.md

```diff
@@ -0,0 +1,243 @@
+- If multiple functions with the same name are defined or a called function is missing, the interpreter throws an error.
+- `main` functions in the text and JSON Bril representations can take arguments that are fed to `brili`. `main` also takes named and typed arguments, rather than C-style `argc/argv`.
+However, `main` doesn't return an exit code for simplicity.
+- Originally, the Bril interpreter simply threw string message exceptions on errors. We made the design decision that the interpreter should not leak interpretation details through uncaught excerptions for anticipated failures. We updated the interpreter to return a specific exception, which is caught and send to standard error along with a custom exit code.
+
+### Challenges
+
+The hardest part of this particular project, as with many compiler endeavors, was wrangling with new frameworks and existing code bases. 
+In particular, this project was more involved than we originally expected because it touched the full Bril stack&mdash;not just the interpreter, but the text-to-JSON and JSON-to-text compilers, the TypeScript frontend, and the Turnt testing framework. 
+
+The TypeScript frontend changes were especially gnarly because the TypeScript AST does not have detailed documentation. It took us quite some time to determine, e.g., if a function call AST node stored its result to a variable. 
+
+Finally, the Hypothesis testing framework was completely new for us, so it was somewhat challenging to think of how to generate meaningful test data automatically. In the end, we settled on generating relatively simple syntactically correct programs. It would be interesting to put more time into generating richer, semantically meaningful Bril in the future as well. 
+
+## Evaluating our contribution
+
+To convince ourselves that we'd actually made a useful contribution to Bril, we wanted to rigorously test our changes. 
+Our evaluation was two-fold: (1) manual testing at multiple abstraction levels (JSON, text-based Bril, and TypeScript), and (2) automated property-based testing to try and cover classes of errors we may not have anticipated. In order to support these lofty testings goals, we also had to make several tooling changes.
+
+### Tooling changes for testing
+
+As we developed our implementation, we built up a  bevy of small Bril programs that we expected to trigger certain classes of errors.
+However, the check-expect-style testing framework Bril employs, [Turnt][], did not support tests that were expected to fail.
+We [extended][] Turnt to check both standard error and program exit codes in order to test invalid Bril programs. 
+
+Turnt relies on C-style comments to configure settings on a per-test basis, so we also extended the Bril text-based surface syntax to support comments of the form `\\ <comment>`. 
```

Comment:
Oops; I also added comments. :flushed:
https://github.com/sampsyo/bril/commit/87c70e98b21bcb50f51fac9a176183a1d2f5a173

Reply:
Ah we can revert ours when we merge, but will keep this in the blog post for now